test/unit/master/test_pagination.py

```diff
@@ -0,0 +1,118 @@
+from typing import Optional
+from threading import Event
+
+from genty import genty, genty_dataset
+from hypothesis import given
+from hypothesis.strategies import text, dictionaries, integers
+from unittest.mock import MagicMock, Mock
+
+from app.master.build import Build
+from app.master.build_request import BuildRequest
+from app.master.cluster_master import ClusterMaster
+from app.slave.cluster_slave import SlaveState
+from app.util.exceptions import BadRequestError, ItemNotFoundError
+from app.web_framework.cluster_base_handler import ClusterBaseHandler, pagination_constants
+from test.framework.base_unit_test_case import BaseUnitTestCase
+
+
+@genty
+class TestClusterMaster(BaseUnitTestCase):
+
+    def setUp(self):
+        super().setUp()
+        self.patch('app.util.fs.create_dir')
+        self.patch('app.util.fs.async_delete')
+        self.patch('os.makedirs')
+        self.mock_slave_allocator = self.patch('app.master.cluster_master.SlaveAllocator').return_value
+        self.mock_scheduler_pool = self.patch('app.master.cluster_master.BuildSchedulerPool').return_value
+
+    @genty_dataset(
+        no_params=(
+            None, None,
+            pagination_constants['DEFAULT_OFFSET'] + 1,
+            pagination_constants['DEFAULT_OFFSET'] + pagination_constants['DEFAULT_LIMIT']
+        ),
+        offset_param=(
+            50, None,
+            50 + 1,
+            50 + pagination_constants['DEFAULT_LIMIT']
+        ),
+        limit_param=(
+            None, 100,
+            pagination_constants['DEFAULT_OFFSET'] + 1,
+            pagination_constants['DEFAULT_OFFSET'] + 100
+        ),
+        offset_and_limit_params=(
+            53, 100,
+            53 + 1,
+            53 + 100
+        ),
+        low_offset=(
+            None, 2,
+            pagination_constants['DEFAULT_OFFSET'] + 1,
+            pagination_constants['DEFAULT_OFFSET'] + 2
+        ),
+        too_high_offset=(
+            1000, 100,
+            pagination_constants['DEFAULT_OFFSET'] + 1,
+            pagination_constants['DEFAULT_OFFSET'] + 100
+        ),
+        too_high_limit=(
+            None, 1000,
+            pagination_constants['DEFAULT_OFFSET'] + 1,
+            pagination_constants['DEFAULT_OFFSET'] + pagination_constants['MAX_LIMIT']
+        ),
+        negative_offset=(
+            -50, None,
+            0 + 1,
+            pagination_constants['DEFAULT_OFFSET'] + pagination_constants['DEFAULT_LIMIT']
+        ),
+        negative_limit=(
+            None, -50,
+            None,
+            None
+        ),
+    )
+    def test_pagination_request_with_query_params(
+            self,
+            offset: Optional[int],
+            limit: Optional[int],
+            expected_first_build_id: int,
+            expected_last_build_id: int,
+            ):
+        master = ClusterMaster()
+        for build_id in range(1, 501):
```

Comment:
Also I don't understand how moving the pagination default values to the config would affect the structure of these particular tests? I do think moving the default pagination values to the config might make more sense, I was wondering where a good place for them to live would be. With them being in the config they'd be more customizable which would be nice.

Are you suggesting these particular tests shouldn't pass any parameters to `builds` but rather only test the default values?